src/Ouroboros/Storage/Immutable/DB.hs

```diff
@@ -282,45 +296,192 @@ modifyInternalState db newEpoch action = do
+                , _it_epoch_index  :: TVar m (U.Vector (Word64, Word64))
+                -- ^ We load in memory the index file for the epoch we are
+                -- currently iterating on, as it's going to be small anyway
+                -- (usually ~150kb).
+                }
+
+type IteratorResult = Maybe (Epoch, RelativeSlot, ByteString)
+
+-- | Steps an 'Iterator, yielding either an 'ImmutableDBError' or some data
+-- alongside with the 'RelativeSlot' and 'Epoch' this data refers to.
+-- Returns Nothing when there is no more data to be fetched.
+-- NOTE(adn): This works under the assumption that the user is the exclusive
+-- owner of the iterator.
+iteratorNext :: forall m. (MonadSTM m, MonadMask m, HasFSE m)
+             => ImmutableDB m
+             -> Iterator m
+             -> m (Either ImmutableDBError IteratorResult)
+iteratorNext db it@Iterator{..} = do
+  current <- atomically $ readTVar _it_current
+  case outOfBounds current _it_end of
+      True  -> runExceptT $ do
+          ExceptT $ iteratorClose it
+          return Nothing
+      False -> do
+          -- Read more data and step the iterator
+          (eHnd, fullIndex) <- atomically $ do
+              e <- readTVar _it_epoch_handle
+              i <- readTVar _it_epoch_index
+              return (e,i)
+          r <- readNext eHnd fullIndex (snd current)
+          runExceptT $ case r of
+              Left err -> liftFsError $ do
+                  hClose eHnd
+                  throwError err
+              Right bs -> do
+                  -- Step the iterator
+                  iteratorStep
+                  return  $ Just (fst current, snd current, bs)
+  where
+    outOfBounds :: (Epoch, RelativeSlot) -> (Epoch, RelativeSlot) -> Bool
+    outOfBounds (currentEpoch, currentSlot) (endEpoch, endSlot)
+      | currentEpoch > endEpoch                           = True
+      | currentEpoch == endEpoch && currentSlot > endSlot = True
+      | otherwise                                         = False
+
+    readNext :: (Monad m, HasFSE m)
+             => FsHandleE m
+             -> U.Vector (Word64, Word64)
+             -> RelativeSlot
+             -> m (Either FsError ByteString)
+    readNext eHnd fullIndex (RelativeSlot slot) = runExceptT $ do
+          -- Step 1: grab from the cached index the blobSize.
+          let (_, !blobSize) = fullIndex U.! (fromIntegral slot)
+
+          -- Step 2: Read from the epoch file. No need for seeking: as we
+          -- are streaming, we are already positioned at the correct place.
+          hGet eHnd (fromEnum blobSize)
+
+    -- Steps the internal state of an 'Iterator'. Internal use only.
+    iteratorStep :: (MonadSTM m, MonadMask m, HasFSE m)
+                 => ExceptT ImmutableDBError m ()
+    iteratorStep = liftFsError $ do
+        ((epoch, RelativeSlot slot), fullIndex) <- atomically $ do
+            (,) <$> readTVar _it_current <*> readTVar _it_epoch_index
```

Comment:
We already read `_it_epoch_index`; let's just pass the index as an argument; doesn't really change anything, just clarifies the code a tiny bit.

Reply:
Same for the handle.